pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,371 @@
+
+	expected := map[cachemanager.CfDataKey]interface{}{
+		cmKey:  nil,
+		podKey: nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+
+	fi.SetFailures("ConfigMapList", 5)
+
+	// this call should schedule a cache wipe and a replay for the configMapGVK
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+
+	expected2 := map[cachemanager.CfDataKey]interface{}{
+		cmKey: nil,
+	}
+	require.Eventually(t, expectedCheck(cfClient, expected2), eventuallyTimeout, eventuallyTicker)
+}
+
+// TestCacheManager_concurrent makes sure that we can add and remove multiple sources
+// from separate go routines and changes to the underlying cache are reflected.
+func TestCacheManager_concurrent(t *testing.T) {
+	r := rand.New(rand.NewSource(12345)) // #nosec G404: Using weak random number generator for determinism between calls
+
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+	testResources, ctx := makeTestResources(t, mgr, wm, c)
+
+	cacheManager := testResources.CacheManager
+	dataStore := testResources.CFDataClient
+	agg := testResources.GVKAgreggator
+
+	// Create configMaps to test for
+	cm := unstructuredFor(configMapGVK, cm1Name)
+	require.NoError(t, c.Create(ctx, cm), fmt.Sprintf("creating ConfigMap %s", cm1Name))
+	t.Cleanup(func() {
+		assert.NoError(t, deleteResource(ctx, c, cm), fmt.Sprintf("deleting resource %s", cm1Name))
+	})
+	cmKey, err := cachemanager.KeyFor(cm)
+	require.NoError(t, err)
+
+	cm2 := unstructuredFor(configMapGVK, cm2Name)
+	require.NoError(t, c.Create(ctx, cm2), fmt.Sprintf("creating ConfigMap %s", cm2Name))
+	t.Cleanup(func() {
+		assert.NoError(t, deleteResource(ctx, c, cm2), fmt.Sprintf("deleting resource %s", cm2Name))
+	})
+	cm2Key, err := cachemanager.KeyFor(cm2)
+	require.NoError(t, err)
+
+	pod := unstructuredFor(podGVK, pod1Name)
+	require.NoError(t, c.Create(ctx, pod), fmt.Sprintf("creating Pod %s", pod1Name))
+	t.Cleanup(func() {
+		assert.NoError(t, deleteResource(ctx, c, pod), fmt.Sprintf("deleting resource %s", pod1Name))
+	})
+	podKey, err := cachemanager.KeyFor(pod)
+	require.NoError(t, err)
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	syncSourceTwo := aggregator.Key{Source: "source_b", ID: "ID_b"}
+
+	wg := &sync.WaitGroup{}
+
+	// simulate a churn-y concurrent access by swapping the gvks for the sync sources repeatedly
+	// and removing sync sources, all from different go routines.
+	for i := 1; i < 100; i++ {
+		wg.Add(3)
+
+		// add some jitter between go func calls
+		time.Sleep(time.Duration(r.Intn(jitterUpperBound)) * time.Millisecond)
+		go func() {
+			defer wg.Done()
+
+			require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
```

Comment:
`require` cannot be used in child goroutines:

https://github.com/stretchr/testify/issues/772